# NumPy Basics
## ENSIAS SURVIVAL KIT
___Author : EZZAIM SOUFIANE___

### Motivation
* __pandas__
* __scikit-learn__
* __tensorflow__
* __pytorch__

### Why np ?
* __ndarray__
* __array oriented arithmeic__
* __vectorized fast math functions__


### How to np ?

In [1]:
import numpy as np

 __ Write a code to__
 
* __sort "this"__
* __ find max of "this"__
* __ find the position of the max in "this"__

## this = [4,7,3,9,5] 


In [3]:
this = np.array([4, 7, 3, 9, 5], dtype=np.float)
print (this)

[4. 7. 3. 9. 5.]


__Write a code to __
  *  __add "this" and "that"__
  *  __find max of this "this" and "that"__
  *  __add "5" with "that" __

## that = [10,20,30,40,50]


In [4]:
that = np.array([10, 20, 30, 40, 50])
this>that

array([False, False, False, False, False])

## Need for Speed

Compare numpy with random module and check the time of execution using time magic

In [27]:
%%timeit
import random
sorted(random.sample(range(int(1e5)),int(1e4)))

19.4 ms ± 195 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [28]:
%%timeit
np.sort(np.random.choice(np.arange(1e5),int(1e4)))

1.5 ms ± 4.45 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


## Getting help from numpy <br>
__What to do if shift+Tab doesnt work!!!__

In [10]:
# np.lookfor("eigen")

### The NumPy ndarray: A Multidimensional Array Object
* __from list__

In [47]:
list_array = np.array([1, 2, 3, 4, 5])
list_array.shape

(5,)

* __from list of list__      <br>Eg: matrix

In [23]:
lol_array = np.array([[[2, 5], 
                      [4, 3],
                      [1, 0],
                      [2, 1]]])

lol_array

array([[[2, 5],
        [4, 3],
        [1, 0],
        [2, 1]]])

* __Basic Details of an Array__  - shape, dimension, type

In [24]:
lol_array.shape

(1, 4, 2)

In [17]:
lol_array.size

8

In [21]:
lol_array.ndim

2

* __other ways of array creation__

Create an array of zeros of shape (2,2)

In [28]:
null = np.zeros((2, 2), dtype=np.int)
null

array([[0, 0],
       [0, 0]])

Create an array of ones of shape (3,4)

In [32]:
unit = np.ones((3, 4))
unit.ndim

2

Create an array of whole numbers ranging from 0 to 9

In [46]:
range_nos = np.arange(2, 9)
range_nos.shape

(7,)

Create an array of even nos between 0 and 11

In [48]:
even_nos = np.arange(0, 11, 2)
even_nos

array([ 0,  2,  4,  6,  8, 10])

Create an array of equally spaced 5 nos between 1 and 2. <br>
Hint: Stepsize is not known, so do not use np.arange() 

In [51]:
interval_split = np.linspace(1, 2, 7)
interval_split

array([1.        , 1.16666667, 1.33333333, 1.5       , 1.66666667,
       1.83333333, 2.        ])

Create an empty array like another array

In [55]:
arr = np.empty((3, 7))

In [52]:
empty_array = np.empty_like(even_nos)
empty_array

array([         0, 1072693248,          0, 1073217536,          0,
       1073741824])

### Arithmetic with NumPy Arrays
__Vectorization__ <br>
__What does it mean? No need for loops__



Create an array 'this' which has the nos 2,10, 12

In [56]:
this = np.array([2, 10, 2])


Divide the elements of the array by 1/2

In [62]:
this/2

array([1., 5., 1.])

Create an array 'that' which has the nos 4,5,6

In [79]:
that = np.array([4, 5, 6])

Compare the two arrays

In [80]:
this>that

array([False,  True, False])

### Doctor Advise: Excercise is Good

* __Create two arrays A and B__

$A = \begin{bmatrix}1 & 2 & 3 \\4 & 5 & 6 \end{bmatrix}$ <br>

$B = \begin{bmatrix}3 & 7 & 2 \\0 & 2 & 4 \end{bmatrix}$ <br>

* Subtract the arrays from each other



In [82]:
A = np.array([[1, 2, 3],
             [4, 5, 6]])

B = np.array([[3, 7, 2],
             [0, 2, 4]])

A-B

array([[-2, -5,  1],
       [ 4,  3,  2]])

### Array Slicing
__sniper not periscope__

Create an array of nos from 0 to 10 and extract the first 5 nos.

In [84]:
arr = np.arange(11)
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [85]:
arr[:5]

array([0, 1, 2, 3, 4])

Substitute the last 4 nos with zeros

In [90]:
arr[-4:] = 0
arr

array([0, 1, 2, 3, 4, 5, 6, 0, 0, 0, 0])

__Never loose the plot__

Extract every no in the even position

In [97]:
arr[::2]

array([0, 2, 4, 6, 0, 0])

__Reverse the array__

In [101]:
arr[::-1]

array([0, 0, 0, 0, 6, 5, 4, 3, 2, 1, 0])

__copy saves__

Create a copy which is not tied to the original array

In [105]:
copy_arr = np.copy(arr)
copy_arr
# copy_arr = arr.copy()
# copy_arr

array([0, 1, 2, 3, 4, 5, 6, 0, 0, 0, 0])

__Get used to 2D Array__

Create a 2D array of nos from 1 to 12 (shape of the array should be (4, 3))

In [122]:
arr = np.arange(1, 13)
print(twod_arr.shape)
new_arr = np.reshape(arr, (4, 3))
print(new_arr, new_arr.shape)

(12,)
[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]] (4, 3)


In [123]:
new_arr[0, 0]

1

### Arithmetic with NumPy Arrays
* 1D array __Broadcasting __ <br>
If two array differ in dimensions, then the shape of the array with fewer dimensions is padded with 1s on the left side.

Create an array of nos from 0 to 3

In [146]:
arr = np.arange(3)
print(arr, arr.shape)

new_arr = np.arange(1, 13)
arr_2 = np.reshape(new_arr, (4, 3))

print(arr_2, arr_2.shape)

[0 1 2] (3,)
[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]] (4, 3)


Add the two arrays and print the shape. What do you infer???

In [147]:
arr + arr_2

array([[ 1,  3,  5],
       [ 4,  6,  8],
       [ 7,  9, 11],
       [10, 12, 14]])

* Multi-dimensional array __Broadcasting __ <br>
If two array differ in dimensions, then numpy tries to stretch the array wherever its shape is 1 to match the other array

Create an array of nos from 0 to 3 of shape (4, 1)

In [169]:
new_arr = np.arange(4)
arr_3 = new_arr[:, np.newaxis]
arr_3.shape

(4, 1)

In [170]:
arr_2*arr_3

array([[ 0,  0,  0],
       [ 4,  5,  6],
       [14, 16, 18],
       [30, 33, 36]])

Multiply the two arrays and print the shape. What do you infer???

## Indexing numpy arrays

Extract the first element of the 2d array

In [172]:
arr = np.reshape(np.arange(0, 15), (3, 5))
arr

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14]])

In [135]:
arr[0, 0]

0

Extract the first two rows of the array

In [130]:
arr[:2, :]

array([10, 11, 12, 13, 14])

Extract the first two columns of the array

In [133]:
arr[:, :2]

array([[ 0,  1],
       [ 5,  6],
       [10, 11]])

Extract the elements in the first two rows and the first two columnst

In [173]:
arr[:]

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14]])

### Doctor Advise: Excercise is Good

* __select the second row but only the first two columns__
* __select the third column but only the first two rows__
* __select all rows of the second column __

$C = \begin{bmatrix}1 & 2 & 3 \\7 & 5 & 8 \\ 10 & 20 & 30\end{bmatrix}$

* __Narrate the following slices:__<br>
__arr [ : , : 2 ]__ <br>
__arr [ 1, 1 : ]__ <br>
__arr [ : ]__

### Bolean Indexing
Numpy arrays can also be indexed by boolean values!! <br>
__ my friends names in MOOC class and their assignment scores__

In [176]:
names = np.array(['ali','bob','chad','daisy'])
scores = np.array([
    [40,45,42],
    [10,15,7],
    [50,50,49],
    [50,2,50]
])
scores

array([[40, 45, 42],
       [10, 15,  7],
       [50, 50, 49],
       [50,  2, 50]])

__Daisy is my friend!! Make her scores 100!!__

Get the scores of daisy

In [193]:
bool_idx = (names == "daisy")
scores[bool_idx]

scores[[False, False, False, True], [True, True, False]]

array([100, 100])

Reset all scores of daisy to 100

In [191]:
scores[bool_idx] = 100
scores

array([[ 40,  45,  42],
       [ 10,  15,   7],
       [ 50,  50,  49],
       [100, 100, 100]])

* __Reshape__

Create an np array of nos from 0 to 31

In [196]:
arr = np.arange(0, 32)
arr.shape

(32,)

Reshape the array to a 4x8 matrix

In [205]:
matrix = np.reshape(arr, (-1, 4))
matrix.shape

(8, 4)

__hey, are you lazy? __  Provide the first index of the new np array and numpy takes care of it.

In [201]:
matrix_1 = np.reshape(arr, (2, -1))
matrix_1.shape

(2, 16)

Bring the array back to original size

In [213]:
orig_arr = np.reshape(matrix_1, (-1))
orig_arr.shape

(32,)

* __How to add a new dimension to an array ??__

In [214]:
with_new_dim = orig_arr[:, np.newaxis]
with_new_dim.shape

(32, 1)

In [216]:
with_new_dim = orig_arr[:, np.newaxis, np.newaxis, np.newaxis]
with_new_dim.shape

(32, 1, 1, 1)

### Fancy Indexing - Revisited

**ATTENTION: The most wrong way of indexing and iterating through an numpy array!!!!**

Task: Create a numpy array of nos from 0 to 19 of shape 5x4. Print the rows of the array

In [219]:
arr = np.arange(20).reshape((5, 4))
print(arr)

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]
 [16 17 18 19]]


**Instead use list of indices!! ** __pick the 1,3 rows__

In [224]:
arr[[1, 3]]

array([[ 4,  5,  6,  7],
       [12, 13, 14, 15]])

__last two rows__

In [231]:
arr[-2:]

array([[12, 13, 14, 15],
       [16, 17, 18, 19]])

__guess the output__

In [232]:
arr[[1,2,3],[3,2,1]]

array([ 7, 10, 13])

__Still, we can make it!__

### Universal Functions: Fast Element-Wise Array Functions
These functions operate on an numpy array in an element wise manner and support broadcasting and type  casting <br>
Ref: https://docs.scipy.org/doc/numpy-1.14.0/reference/ufuncs.html

Create an array of nos from 0 to 10

In [243]:
arr = np.arange(1, 11)

__unary ufuncs__

Print the square and sqrt of the elements of the numpy array

In [244]:
np.square(arr[2:6])

array([ 9, 16, 25, 36], dtype=int32)

In [245]:
np.sqrt(arr)

array([1.        , 1.41421356, 1.73205081, 2.        , 2.23606798,
       2.44948974, 2.64575131, 2.82842712, 3.        , 3.16227766])

Print the exponential of an array

In [246]:
np.exp(arr)

array([2.71828183e+00, 7.38905610e+00, 2.00855369e+01, 5.45981500e+01,
       1.48413159e+02, 4.03428793e+02, 1.09663316e+03, 2.98095799e+03,
       8.10308393e+03, 2.20264658e+04])

Print the reciprocal of the elements of an array

In [247]:
np.reciprocal(arr)

array([1, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

__binary ufuncs__

Find the quotient and remainder

In [249]:
quotient, remainder = np.divmod(arr, 5)

Check if two arrays are equal

In [251]:
# quotient == remainder        Use either way
np.equal(quotient, remainder)

array([False, False, False, False, False,  True, False, False, False,
       False])

Get the maximum of the two arrays

In [254]:
# np.greater(quotient, remainder)
quotient > remainder

array([False, False, False, False,  True, False, False, False, False,
        True])

### Doctor Again! Excercise Again!

* date = ['01/01/2018','01/02/2018','01/03/2018','01/04/2018','01/05/2018'] <br\>
* stock_price = ['44','47.9','51','48.5','53']

#### quants suggested:
    * sell if price goes above 50 - Mark the output array with 'SELL'
    * buy if price goes less than 45 - Mark the output array with 'BUY'
    * hold if price is between 45 and 50 - Mark the output array with 'HOLD'
    
#### expected output: 
    * position : ['HOLD', 'BUY', 'SELL', 'BUY', 'SELL']

In [23]:
date = ['01/01/2018','01/02/2018','01/03/2018','01/04/2018','01/05/2018']
stock_price = ['44','47.9','51','48.5','53']

#convert
date_ = np.array(date)
price_ = np.array(stock_price,dtype=np.float)

#compute
sell = np.greater(price_, 50)
hold = np.less(price_, 45)
buy = np.logical_and(np.greater_equal(price_, 45), np.less_equal(price_, 50))
#buy = (price_ >= 45) & (price_ <= 50)

#set
position = np.empty_like(date_)
position[sell] = 'SELL'
position[hold] = 'HOLD'
position[buy] = 'BUY'
position


array(['HOLD', 'BUY', 'SELL', 'BUY', 'SELL'], dtype='<U10')

__when to hold?__

In [24]:
date_[position == 'HOLD']

array(['01/01/2018'], dtype='<U10')

### Array-Oriented Programming with Arrays
__for-in-if __

Wherever the price is greater than 50 put 'SELL' otherwise replace with 'BUY'

**Not the right way to do it!!!**

In [25]:
['SELL' if float(p) > 50 else 'BUY' for p in stock_price]


['BUY', 'BUY', 'SELL', 'BUY', 'SELL']

**The right way!!!**

In [26]:
np.where(price_ >50,'SELL','BUY')

array(['BUY', 'BUY', 'SELL', 'BUY', 'SELL'], dtype='<U4')

### ML
#### Squared euclidean distance 
__$ (x1-x2)^2 + (y1-y2)^2 $__

In [255]:
vector1 = np.array([1,2])
vector2 = np.array([3,2])


The Pythonic way!! Not to be used with numpy arrays

In [262]:
sum([(x-y)**2 for x, y in zip(vector1, vector2)])

4

The Numpy way!!

In [263]:
np.sum(np.power(np.subtract(vector1, vector2), 2))

4

Another easy way!!

In [264]:
((vector1 - vector2)**2).sum()

4

### statistics

Generate an array of 4 random integers between 1 and 7

In [268]:
rand_arr = np.random.randint(1, 8, size=4)
rand_arr

array([3, 5, 7, 1])

Find the mean of the array

In [269]:
np.mean(rand_arr)

4.0

Find the cummulative sum of elements

In [270]:
np.cumsum(rand_arr)

array([ 3,  8, 15, 16], dtype=int32)

Find the maximum element and its position.

In [274]:
print(np.amax(rand_arr), np.argmax(rand_arr))
# np.amin, np.argmin

7 2


Generate an array of 10 random number from the normal distribution.

In [278]:
normal_arr = np.random.randn(10)
normal_arr

array([ 0.14269117, -1.35570392, -0.4579673 , -2.30287656,  0.66694298,
       -0.38043054,  1.06413683, -0.17405051,  0.46008006, -0.67496334])

Find the number of elements greater than 0

In [281]:
normal_arr[normal_arr>0]         # I got non zero elements

# Position of non-zero elements
np.where(normal_arr>0)

(array([0, 4, 6, 8], dtype=int64),)

### Sorting

Create a randomly permutated array of range 12 and shape 4,3

In [286]:
# arr = np.arange(13)
# print(arr)
arr = np.random.permutation(12).reshape(4, 3)
arr

array([[ 3,  8,  7],
       [ 4,  2,  6],
       [ 0,  9,  1],
       [11,  5, 10]])

Sort the elements of the array along columns

In [292]:
np.sort(arr, axis=0)  # VERTICAL

array([[ 0,  2,  1],
       [ 3,  5,  6],
       [ 4,  8,  7],
       [11,  9, 10]])

Sort the elements of the array along rows

In [293]:
np.sort(arr, axis=1)  # HORIZONTAL

array([[ 3,  7,  8],
       [ 2,  4,  6],
       [ 0,  1,  9],
       [ 5, 10, 11]])

### set in Numpy

In [298]:
names = np.array(['Bob', 'Joe', 'Will', 'Bob', 'Will', 'Joe', 'Joe'])


Find the unique names in the array

In [299]:
np.unique(names)

array(['Bob', 'Joe', 'Will'], dtype='<U4')

### Linear Algebra

Create a numpy array of the matrix $A = \begin{bmatrix}1 & 2 & 3 \\4 & 5 & 6 \end{bmatrix}$

In [3]:
A = np.array([[1, 2, 3],
             [4, 5, 6]])
A

array([[1, 2, 3],
       [4, 5, 6]])

Create a numpy array of the matrix $B = \begin{bmatrix}2 & 2 \\1 & 1\end{bmatrix}$

In [4]:
B = np.array([[2, 2], [1, 1]])
B

array([[2, 2],
       [1, 1]])

Multiply the two matrices

In [6]:
C = np.dot(B, A)
C

array([[10, 14, 18],
       [ 5,  7,  9]])

Another way to multiply

In [8]:
C = B@A
C

array([[10, 14, 18],
       [ 5,  7,  9]])

Create a numpy array of the matrix $C = \begin{bmatrix}1 & 2 & 3 \\7 & 5 & 8 \end{bmatrix}$

In [10]:
C = np.array([[1, 2, 3], [7, 5, 8]])
C

array([[1, 2, 3],
       [7, 5, 8]])

Find the element-wise product (vector dot product or inner product) of A and C

In [11]:
E = A*C
E

array([[ 1,  4,  9],
       [28, 25, 48]])

Find the transpose of the matrix

In [13]:
C.T             # Shorthand
np.transpose(C)

array([[1, 7],
       [2, 5],
       [3, 8]])

Find the determinant of the matrix

In [15]:
np.linalg.det(B)

0.0

Find the inverse of the matrix

In [17]:
np.linalg.pinv(B)

array([[0.2, 0.1],
       [0.2, 0.1]])

Create an identity matrix of size 5x5

In [18]:
np.identity(5)

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

Extract the diagonal elements of matrix B

In [19]:
np.diag(B)

array([2, 1])

Create a diagonal matrix of the 1d array [3, 4, 5]

In [20]:
arr = np.array([3, 4, 5])
np.diag(arr)

array([[3, 0, 0],
       [0, 4, 0],
       [0, 0, 5]])

#### LinAlg library in Numpy

**a)** Create the matrix $A = \begin{pmatrix} -2 & -4 & 2\\ -2 & 1 & 2 \\ 4 & 2 & 5\end{pmatrix}$

**b)** Calculate the eigenvalues and eigenvectors. Print each eigenvalue with the corresponding eigenvector

**c)** Create the matrix $B = \begin{pmatrix} 2 & 2 \\ 5 & -1 \end{pmatrix}$

**d)** Calculate the eigenvalues and eigenvectors of $B$

**e)** Calculate the length of the vector $y$ = $B*x$, where $x = \begin{pmatrix} \frac{1}{\sqrt{2}} \\ \frac{1}{\sqrt{2}} \end{pmatrix}$ (use the function *norm* from the *linalg* package). What do you notice?

In [3]:
A = np.array([[-2, -4, 2],
             [-2, 1, 2],
             [4, 2, 5]])

In [6]:
eigval, eigvec = np.linalg.eig(A)

for value, vec in zip(eigval, eigvec):
    print("Eifen value {}     Eigen vector {}".format(value, vec))

Eifen value -5.000000000000005     Eigen vector [0.81649658 0.53452248 0.05842062]
Eifen value 2.999999999999998     Eigen vector [ 0.40824829 -0.80178373  0.35052374]
Eifen value 5.999999999999999     Eigen vector [-0.40824829 -0.26726124  0.93472998]


In [7]:
B = np.array([[2, 2],
             [5, -1]])

In [8]:
eigval, eigvec = np.linalg.eig(B)

for value, vec in zip(eigval, eigvec):
    print("Eifen value {}     Eigen vector {}".format(value, vec))

Eifen value 4.0     Eigen vector [ 0.70710678 -0.37139068]
Eifen value -3.0     Eigen vector [0.70710678 0.92847669]


In [12]:
x = np.array([[1/np.sqrt(2)], [1/np.sqrt(2)]])
np.linalg.norm(B*x)

4.12310562561766

Calculate the determinant of the matrix $A = \begin{pmatrix} 1 & 2 & 3 \\ 0 & 4 & 5 \\ 1 & 0 & 6 \end{pmatrix}$



In [13]:
A = np.array([[1, 2, 3],
             [0, 4, 5],
             [1, 0, 6]])

Calculate the inverse of $A$. Verify by calculating $A * A^{-1}$. You should get the identity matrix.



In [15]:
A_inv = np.linalg.inv(A)
np.dot(A, A_inv)

array([[ 1.00000000e+00,  2.77555756e-17,  5.55111512e-17],
       [ 5.55111512e-17,  1.00000000e+00, -5.55111512e-17],
       [-1.11022302e-16,  5.55111512e-17,  1.00000000e+00]])

Consider the matrix $B = \begin{pmatrix} 1 & 2 & 3 \\ 4 & 5 & 6 \\ 5 & 7 & 9\end{pmatrix}$. Is this matrix invertible?

In [22]:
B = np.array([[1, 2, 3],
             [4, 5, 6],
             [5, 7, 9]])
np.linalg.inv(B)

array([[ 2.04709074e+15,  2.04709074e+15, -2.04709074e+15],
       [-4.09418148e+15, -4.09418148e+15,  4.09418148e+15],
       [ 2.04709074e+15,  2.04709074e+15, -2.04709074e+15]])

#### Distributions


Generate an array of 10 random nos sampled from the uniform distribution

In [23]:
np.random.rand(10)

array([0.32147224, 0.61243099, 0.1393467 , 0.48720037, 0.92585671,
       0.2204819 , 0.90399446, 0.27992905, 0.19037707, 0.79958592])

Generate an array of size (2,2) sampled from the normal distribution

In [29]:
np.random.randn(2, 2)

array([[ 0.56960292,  0.40509519],
       [-2.82320466,  0.64209672]])

Generate an array of 10 random nos sampled from the normal distribution with mean 2 and std 4

In [30]:
np.random.normal(loc=2, scale=4, size=(2, 2))

array([[-6.57574531,  2.64699621],
       [-0.61056313, -1.5041835 ]])

Generate an array of size (3,4) with nos sampled from the uniform distribution between 5 and 10

In [31]:
np.random.uniform(low=5, high=11, size=(3,4))

array([[ 7.55522769,  9.3466096 ,  7.52513184,  5.82706178],
       [ 6.94598268,  7.67381096,  7.46695855,  9.98241038],
       [ 8.77547848,  7.94284128,  7.46758592, 10.25484406]])

Generate an array of size (2,3) with nos sampled from the discrete uniform distribution between 1 and 8

In [32]:
np.random.randint(low=1, high=9, size=(2, 3))

array([[8, 5, 5],
       [7, 1, 4]])

### Further reading:
* Numpy docs
* Numpy 1.5 Beginners Guide
* Guide to Numpy http://web.mit.edu/dvp/Public/numpybook.pdf
* 100 Numpy exercises http://www.labri.fr/perso/nrougier/teaching/numpy.100/